### ISSUE TITLE
[resize2fs]  lots of warnings when expanding very small ext4 fs to 1.5TB

### ISSUE THREAD

**Issue by ralisi** [tytso/e2fsprogs#60]
Hi,

I have come along doing the same weirdnesses for the very same reasons as described in #50.

Doing the following commands on ea82add307c4bc820423abc8acc2f155720cf914 gives a lot of warnings when resizing:
`Illegal block number passed to ext2fs_test_block_bitmap #12440 for block bitmap for broken.img`

```
rm -f broken.img
truncate -s 10M broken.img
./misc/mke2fs broken.img -b 1024 -T default
truncate -s +1500G broken.img
./resize/resize2fs broken.img
```

As stated in #50, it should not be possible to extend the filesystem past 1023GB, but it seems to work (apart from these warnings) and `fsck` can't find any problem later on. `resize2fs` only refuses to run when I extend past 2TB.

So I am either expecting that my extension should have also been rejected or these warnings should not exist.

**Comment by tytso** [comment]
You're right that resize2fs should either not issue the warnings, or reject the resize.  From that perspective, thank you for the bug report.  I will look into this and try to figure out what is going on, and address this.

Fundamentally, though, the fix from your perspective is the same as #50.   "Doctor, doctor, it hurts when I do that.  Well, then don't do that!"   Trying to resize a file system from super-tiny to super-large is never going to end well.   Even when it "works", the performance is going to be terrible.   That's because mke2fs will be making changes based on hueristics based on the size of the file system.  When you then blow it up by a thousand of times, or more, those hueristics will be not be correct.

In this case, with a small file system, we use a 1k block size, to reduce space wasted by internal fragmentation.   We also reduce the size of the ext4 journal, to reduce file system overhead.   But both of these changes have significant performance impacts --- which might not matter if the file system is meant for a small, slow USB thumb drive, but if you are trying to use this for a higher performance system ---- you'll be sorry.

We also don't enable 64-bit block numbers, since this increases metadata overhead.   But like the 1k block size, this limits the maximuim size that is supported by the base file system, and it is not at all simple to change the fundamental aspects of the file system so that the file system is (a) scalable, and (b) performant.    So this idea of "drop a small file system on a device" and then blowing it up to a huge size is a Really, REALLY, **REALLY** bad idea.    Is there some place which is recommending this anti-pattern, or otherwise causing people to think this is a good idea? 

There are some ways that this can be mitigated (you can give options to mke2fs to use a 4k block size, enable 64k blocksize, and force a larger journal size), but if the file system has some files placed in the small file system, the block placement of the small file system will not be optimal after the file system is blown up to a large size.    So the bug reporters for #50 and #52 are getting the same (bad) idea from some common source, it would be good if we could fix, or remove, the place which is giving bad advice.

**Comment by ralisi** [comment]
Sorry for the very long delay, but I was trying to get to the root cause of this:

> So this idea of "drop a small file system on a device" and then blowing it up to a huge size is a Really, REALLY, REALLY bad idea. Is there some place which is recommending this anti-pattern, or otherwise causing people to think this is a good idea?

It appears to be some general pattern in embedded systems world, for which I could not find an original source, e.g. used by raspian as well, where it already appeared in the first commit from 2012. In that case, however, only intended to run on an SD card. The idea of extending it to 2TB just comes from using the same approach on bigger hardware.
